### ISSUE TITLE
[dynamo] Diffusers - Graph break on OrderedDict 

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#102878]
~~~
import diffusers
from diffusers.utils import BaseOutput
import dataclasses
from dataclasses import dataclass
import torch

@dataclass
class Transformer2DModelOutput(BaseOutput):
    """
    Args:
        sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)` or `(batch size, num_vector_embeds - 1, num_latent_pixels)` if [`Transformer2DModel`] is discrete):
            Hidden states conditioned on `encoder_hidden_states` input. If discrete, returns probability distributions
            for the unnoised latent pixels.
    """

    sample: torch.FloatTensor



def fn(x):
    return Transformer2DModelOutput(sample=x)


x = torch.randn(4)
ref = fn(x)

opt_fn = torch.compile(fn, backend="eager", fullgraph=True)
res = opt_fn(x)
~~~

**Comment by anijain2305** [comment]
A simpler version

~~~
from diffusers.utils import BaseOutput
from collections import OrderedDict
from dataclasses import dataclass
import torch

@dataclass
class Transformer2DModelOutput(OrderedDict):
    sample: torch.FloatTensor



def fn(x):
    return Transformer2DModelOutput(sample=x)


x = torch.randn(4)
ref = fn(x)

opt_fn = torch.compile(fn, backend="eager", fullgraph=True)
res = opt_fn(x)

~~~

**Comment by haofanwang** [comment]
The example above still fails.

**Comment by desertfire** [comment]
No, I haven't done any follow up on this. Is this a blocker for you?

**Comment by haofanwang** [comment]
Yes, I'm also working with diffusers for speeding up and get blocked as the graph breaks. I thought you had already solved it with some commits.

**Comment by desertfire** [comment]
@yanboliang , assigned to you but feel free to reassign to anybody who may have a bandwidth.

**Comment by stroxler** [comment]
The simple second bug (failing on custom dict constructed with kwargs) should be fixed in https://github.com/pytorch/pytorch/pull/112408.

But this won't fix the original bug using BaseOutput, because the original bug isn't actually a problem with handling kwargs to a dict but rather an issue with whether the various VariableTrackers say they match the class.

In more detail, the reason this bugfix won't help with `diffusers.BaseOutput` is that:
- for ModelOutput today, we are using dataclass logic to deal with kwargs, so the fix of kwargs is actually irrelevant
- the underlying problem when inheriting from `diffusers.BaseModel` is that we aren't handling it the same way that we handle ModelOutput, because we've allowlisted special behaviors for only `ModelOutput`

**What to do about it?**

The fix isn't 100% obvious. Allowing `BaseModel` in the same manner that we allow `ModelOutput` is messy, since we need a protected lazy import of `diffusers`. And if we decide to test it the tests will also need to import `diffusers`.

I can put up a PR doing it this way this week, but I'm not sure we really want that approach, especially if we think there might be more cases like this in the future.

**Additional confusion about existing `ModelOutput` behavior**

In addition, the current logic is a little confusing to me (as a dynamo newbie) because we are using a strange mix of `DataClassVariable` and `CustomizedDictVariable` to model `ModelOutput` subclasses today and I'm not convinced this was intentional:
- we currently always use `CustomizedDictVariable` when a compiled function *constructs* a value of one of the classes defined in `transformers.modeling_outputs`
- On the other hand we use a `DataClassVariable` whenever a parameter is passed into a compiled funtion *or* when an end-user-defined subclass` of `ModelOutput` (as opposed to one defined in `transformers.modeling_outputs) is used.
- To me the discussion in https://github.com/pytorch/pytorch/pull/105044/files suggests that we may have intended to always use `CustomizedDictVariable` but there are still a few control flows that don't do this.

**Comment by stroxler** [comment]
Follow-up: it looks like there's actually fix for the `diffusers` case in
https://github.com/pytorch/pytorch/commit/b97afc401843aae2b07dffca8f00d212ba1d470e

I'm not 100% sure it will work as expected if there's an end-user defined subclass (as opposed to one defined inside of `diffusers` or `transformers`) but it should at least result in `ModelOutput` and `BaseOutput` behaving the same
